# Market Association Analysis

In [6]:
df_price=pd.read_csv("Price.csv")
df_price.head()

ProductID                     ProductName           Brand  \
0        1.0           Fairy Platinum Tablet           Fairy   
1        2.0             Komili Riviera Yağı          Komili   
2        3.0     Marmara Birlik Siyah Zeytin  Marmara Birlik   
3        4.0  CityFarm Organik 10'lu Yumurta       City Farm   
4        5.0             Sütaş Pastörize Süt           Sütaş   

   Category_Level_0 Category_Level_1           Category_Level_2  \
0         Deterjan    Bulaşık Yıkama  Bulaşık Makinesi Ürünleri   
1   Gıda, Şekerleme      Sıvı Yağlar                 Zeytinyağı   
2  Süt, Kahvaltılık           Zeytin               Siyah Zeytin   
3  Süt, Kahvaltılık          Yumurta               Özel Yumurta   
4  Süt, Kahvaltılık              Süt              Pastörize Süt   

     Category_Level_3 Fiyat  
0  Tablet Deterjanlar    60  
1  Riviera Zeytinyağı    40  
2   Özel Siyah Zeytin    20  
3        Özel Yumurta    20  
4  Pastörize Sade Süt     5

In [7]:
df_invoice=pd.read_csv("Invoice.csv")
df_invoice.head()

Invoice ID Invoice Date   Invoice_Total Payment Method
0           1    23.03.2019            150          Debit
1           2    23.03.2019             65    Credit Card
2           3    23.03.2019            140          Debit
3           4    23.03.2019             75    Credit Card
4           5    23.03.2019             55    Credit Card

In [8]:
df_invoice_details=pd.read_csv("InvoiceDetails.csv")
df_invoice_details.head()

InvoiceDetailID  InvoiceID  ProductID  Total  Adet
0                1          1          1     60   1.0
1                2          1          7     50   2.0
2                3          1          2     40   1.0
3                4          2          2     40   1.0
4                5          2          4     20   1.0

In [19]:
df=pd.merge(df_invoice_details, df_price, how='left', on=['ProductID']).\
drop(["Brand","Category_Level_0","Category_Level_1","Category_Level_2","Category_Level_3","Fiyat"],axis=1)
df.head(5)

InvoiceDetailID  InvoiceID ProductID  Total  Adet  \
0                1          1         1     60   1.0   
1                2          1         7     50   2.0   
2                3          1         2     40   1.0   
3                4          2         2     40   1.0   
4                5          2         4     20   1.0   

                                       ProductName  
0                            Fairy Platinum Tablet  
1  Prima Premium Care İkiz Paket 5 No Junior 25'Li  
2                              Komili Riviera Yağı  
3                              Komili Riviera Yağı  
4                   CityFarm Organik 10'lu Yumurta

In [20]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [21]:
basket = (df.groupby(['InvoiceID', 'ProductName'])['Adet']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceID'))

In [22]:
basket.head()

ProductName  Beypazarı Maden Suyu 6X200 Ml  Bonne Maman Çilek Reçeli 370 G  \
InvoiceID                                                                    
1                                      0.0                             0.0   
2                                      0.0                             0.0   
3                                      0.0                             0.0   
4                                      1.0                             0.0   
5                                      0.0                             0.0   

ProductName  CityFarm Organik 10'lu Yumurta  Coca-Cola Şekersiz 6X250ML Kutu  \
InvoiceID                                                                      
1                                       0.0                              0.0   
2                                       1.0                              0.0   
3                                       0.0                              0.0   
4                                       2.0                              0.0   
5                                       0.0                              1.0   

ProductName  Doğuş Toz Şeker 1 Kg  Elma Fuji Kg  Fairy Platinum Tablet  \
InvoiceID                                                                
1                             0.0           0.0                    1.0   
2                             0.0           0.0                    0.0   
3                             0.0           0.0                    1.0   
4                             0.0           0.0                    0.0   
5                             0.0           3.0                    0.0   

ProductName  Komili Riviera Yağı  Limon Nostalji File Kg  \
InvoiceID                                                  
1                            1.0                     0.0   
2                            1.0                     0.0   
3                            2.0                     0.0   
4                            0.0                     0.0   
5                            1.0                     0.0   

ProductName  Marmara Birlik Siyah Zeytin    ...     Nutella 750 G  \
InvoiceID                                   ...                     
1                                    0.0    ...               0.0   
2                                    0.0    ...               0.0   
3                                    1.0    ...               1.0   
4                                    0.0    ...               0.0   
5                                    0.0    ...               0.0   

ProductName  Prima Premium Care İkiz Paket 5 No Junior 25'Li  \
InvoiceID                                                      
1                                                        2.0   
2                                                        0.0   
3                                                        0.0   
4                                                        0.0   
5                                                        0.0   

ProductName  Selpak Lavanta Tuvalet Kağıdı 32'li  Sütaş Pastörize Süt  \
InvoiceID                                                               
1                                            0.0                  0.0   
2                                            0.0                  2.0   
3                                            0.0                  0.0   
4                                            1.0                  0.0   
5                                            0.0                  0.0   

ProductName  Sütaş Tava Yoğurt 1000 G  Torku Tereyağ 500 G  \
InvoiceID                                                    
1                                 0.0                  0.0   
2                                 0.0                  0.0   
3                                 0.0                  0.0   
4                                 0.0                  0.0   
5                                 0.0                  0.0   

ProductName  Yayla Dermason Fasulye 1 Kg  Yayla Koçbaşı Nohut 1000 G  \


In [14]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

basket_sets = basket.applymap(encode_units)

In [15]:
basket_sets.head()

ProductName  Beypazarı Maden Suyu 6X200 Ml  Bonne Maman Çilek Reçeli 370 G  \
InvoiceID                                                                    
1                                        0                               0   
2                                        0                               0   
3                                        0                               0   
4                                        1                               0   
5                                        0                               0   

ProductName  CityFarm Organik 10'lu Yumurta  Coca-Cola Şekersiz 6X250ML Kutu  \
InvoiceID                                                                      
1                                         0                                0   
2                                         1                                0   
3                                         0                                0   
4                                         1                                0   
5                                         0                                1   

ProductName  Doğuş Toz Şeker 1 Kg  Elma Fuji Kg  Fairy Platinum Tablet  \
InvoiceID                                                                
1                               0             0                      1   
2                               0             0                      0   
3                               0             0                      1   
4                               0             0                      0   
5                               0             1                      0   

ProductName  Komili Riviera Yağı  Limon Nostalji File Kg  \
InvoiceID                                                  
1                              1                       0   
2                              1                       0   
3                              1                       0   
4                              0                       0   
5                              1                       0   

ProductName  Marmara Birlik Siyah Zeytin    ...     Nutella 750 G  \
InvoiceID                                   ...                     
1                                      0    ...                 0   
2                                      0    ...                 0   
3                                      1    ...                 1   
4                                      0    ...                 0   
5                                      0    ...                 0   

ProductName  Prima Premium Care İkiz Paket 5 No Junior 25'Li  \
InvoiceID                                                      
1                                                          1   
2                                                          0   
3                                                          0   
4                                                          0   
5                                                          0   

ProductName  Selpak Lavanta Tuvalet Kağıdı 32'li  Sütaş Pastörize Süt  \
InvoiceID                                                               
1                                              0                    0   
2                                              0                    1   
3                                              0                    0   
4                                              1                    0   
5                                              0                    0   

ProductName  Sütaş Tava Yoğurt 1000 G  Torku Tereyağ 500 G  \
InvoiceID                                                    
1                                   0                    0   
2                                   0                    0   
3                                   0                    0   
4                                   0                    0   
5                                   0                    0   

ProductName  Yayla Dermason Fasulye 1 Kg  Yayla Koçbaşı Nohut 1000 G  \


# min_support değerini 0.1 (%10) vererek apriori algoritması için başlangıç değerlerini set ediyoruz.

In [16]:
frequent_itemsets = apriori(basket_sets, min_support=0.1, use_colnames=True)

In [17]:
frequent_itemsets

support                                           itemsets
0      0.26                    (Beypazarı Maden Suyu 6X200 Ml)
1      0.32                   (CityFarm Organik 10'lu Yumurta)
2      0.10                             (Doğuş Toz Şeker 1 Kg)
3      0.24                                     (Elma Fuji Kg)
4      0.30                            (Fairy Platinum Tablet)
5      0.34                              (Komili Riviera Yağı)
6      0.32                      (Marmara Birlik Siyah Zeytin)
7      0.16                                    (Nutella 750 G)
8      0.20  (Prima Premium Care İkiz Paket 5 No Junior 25'Li)
9      0.24              (Selpak Lavanta Tuvalet Kağıdı 32'li)
10     0.14                              (Sütaş Pastörize Süt)
11     0.16                         (Sütaş Tava Yoğurt 1000 G)
12     0.18                                         (Çilek Kg)
13     0.12      (Beypazarı Maden Suyu 6X200 Ml, Elma Fuji Kg)
14     0.10  (CityFarm Organik 10'lu Yumurta, Fairy Platinu...
15     0.12  (Komili Riviera Yağı, CityFarm Organik 10'lu Y...
16     0.14  (Marmara Birlik Siyah Zeytin, CityFarm Organik...
17     0.14  (CityFarm Organik 10'lu Yumurta, Selpak Lavant...
18     0.14       (Komili Riviera Yağı, Fairy Platinum Tablet)
19     0.16  (Marmara Birlik Siyah Zeytin, Fairy Platinum T...
20     0.10  (Fairy Platinum Tablet, Selpak Lavanta Tuvalet...
21     0.16  (Marmara Birlik Siyah Zeytin, Komili Riviera Y...
22     0.12               (Komili Riviera Yağı, Nutella 750 G)
23     0.14  (Marmara Birlik Siyah Zeytin, Selpak Lavanta T...

# Destek (Support) Değeri: Bir ilişkinin tüm alışverişler içinde hangi oranda tekrarlandığını belirtir.

# Güven (Confidence) Değeri: X ürününü alan müşterilerin Y ürününü alma olasılığını ortaya belirtir

Güven (X -> Y) = Frekans(X, Y) / Frekans(X)

# Örneğin X ve Y ürünleri için Destek değerinin %20’i ve Güven değerinin %70 olduğunu varsayarsak; Analiz edilen tüm alışverişlerin %20’sinde X ile Y ürünlerinin birlikte geçtiği, X ürününü satın alan müşterilerin %70’inin aynı alışverişte Y ürününü de satın aldığını gösterir.

# Lift Değeri (Lift Value): Kuralın ilginçliği/farklılığını ifade eder. Hesaplaması aşağıdaki gibidir.

Lift Değeri: Support (X, Y) / ( Support(X) * Support(Y) ) 
Lift Değeri: Frekans (X, Y) / ( Frekans(x) * Frekans(Y) )

# Antecedent (LHS: Left Hand Side): Kuralın sol tarafını ifade eder.
# Consequent (RHS: Right Hand Side): Kuralın sağ tarafını ifade eder.

In [18]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.3)
rules = rules.sort_values(['confidence'], ascending=False)
rules.head()

antecedents                       consequents  \
19                        (Nutella 750 G)             (Komili Riviera Yağı)   
21  (Selpak Lavanta Tuvalet Kağıdı 32'li)     (Marmara Birlik Siyah Zeytin)   
9   (Selpak Lavanta Tuvalet Kağıdı 32'li)  (CityFarm Organik 10'lu Yumurta)   
13                (Fairy Platinum Tablet)     (Marmara Birlik Siyah Zeytin)   
16          (Marmara Birlik Siyah Zeytin)             (Komili Riviera Yağı)   

    antecedent support  consequent support  support  confidence      lift  \
19                0.16                0.34     0.12    0.750000  2.205882   
21                0.24                0.32     0.14    0.583333  1.822917   
9                 0.24                0.32     0.14    0.583333  1.822917   
13                0.30                0.32     0.16    0.533333  1.666667   
16                0.32                0.34     0.16    0.500000  1.470588   

    leverage  conviction  
19    0.0656    2.640000  
21    0.0632    1.632000  
9     0.0632    1.632000  
13    0.0640    1.457143  
16    0.0512    1.320000

# İlgilendiğimiz metriğe göre (confidence, lift, conviction ve vd.) Association Rules tablosunu oluşturuyoruz. Burada metric olarak Confidence ve değerini 0.3 (%30) seçtik.

# Sonuçlar
(Nutella 750 G) ve (Komili Riviera Yağı) ürünlerinin birlikte görülme olasılığı (support) %12 (0.12) olduğunu,
(Nutella 750 G) satın alan kişilerin (confidence) %75 (0.75) olasılıkla  (Komili Riviera Yağı)’ınıda satın aldığını,
(Nutella 750 G)'nın yer aldığı alışveriş sepetlerinde (Komili Riviera Yağı)’ının satışının (lift) 2,205882 kat arttığı,
(Nutella 750 G) ve (Komili Riviera Yağı) ürünlerinin birlikte satın alınmalarının birbirlerinden bağımsız olarak satın alınmalarından ne kadar fazla (leverage) 0,0656 olduğunu,
(Nutella 750 G) ve (Komili Riviera Yağı) ürünlerinin birbirleri ile ilişkili (conviction) 2,64 değeri ile olduğunu  söyleyebiliriz.